### PR TITLE
Fix recursive use of parameterized derived types (issue #283)

### PR DESCRIPTION
**Pull request by klausler**

### CODE REVIEW

File: lib/semantics/resolve-names.cc [hunk 1]
```diff
@@ -2809,13 +2811,18 @@ void DeclarationVisitor::Post(const parser::DerivedTypeSpec &x) {
...
+// in this Pre() routine so that recursive use of the derived type can be
+// supported in the components.
+bool DeclarationVisitor::Pre(const parser::DerivedTypeDef &x) {
+  Walk(std::get<parser::Statement<parser::DerivedTypeStmt>>(x.t));
+  Walk(std::get<std::list<parser::Statement<parser::TypeParamDefStmt>>>(x.t));
   auto &scope{currScope()};
   CHECK(scope.symbol() != nullptr);
   CHECK(scope.symbol()->scope() == &scope);
   auto &details{scope.symbol()->get<DerivedTypeDetails>()};
   auto &stmt{std::get<parser::Statement<parser::DerivedTypeStmt>>(x.t)};
```
Comment (tskeith) [inline lib/semantics/resolve-names.cc]: If this declaration were moved up it could be used in line 2818.

File: lib/semantics/resolve-names.cc [hunk 2]
```diff
@@ -2848,39 +2856,45 @@ void DeclarationVisitor::Post(const parser::DerivedTypeDef &x) {
...
 }
 bool DeclarationVisitor::Pre(const parser::DerivedTypeStmt &x) {
   return BeginAttrs();
 }
 void DeclarationVisitor::Post(const parser::DerivedTypeStmt &x) {
+  // Resolve the EXTENDS() clause before creating the derived
+  // type's symbol to foil attempts to recursively extend a type.
+  auto *extendsName{derivedTypeInfo_.extends};
+  const Symbol *extendsType{nullptr};
+  if (extendsName != nullptr) {
```
Comment (tskeith) [inline lib/semantics/resolve-names.cc]: Can we check here for extendsName being the same as the derived type name and provide a better error, like "derived type cannot extend itself"?

File: lib/semantics/resolve-names.cc (hunk 1 above)
Comment (klausler) [reply lib/semantics/resolve-names.cc]: Thanks.

File: lib/semantics/resolve-names.cc (hunk 2 above)
Comment (klausler) [reply lib/semantics/resolve-names.cc]: Sure.

**Pull request merged**